pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type policyWatcher struct {
+	log           logrus.FieldLogger
+	config        *option.DaemonConfig
+	policyManager PolicyManager
+	readStatus    DirectoryWatcherReadStatus
+	// maps cnp file name to cnp object. this is required to retrieve data during delete.
+	fileNameToCnpCache map[string]*cilium_v2.CiliumNetworkPolicy
+}
+
+func (p *policyWatcher) translateToCNPObject(file string) (*cilium_v2.CiliumNetworkPolicy, error) {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	// yaml to json conversion
+	jsonData, err := yaml.YAMLToJSON([]byte(data))
+	if err != nil {
+		return nil, err
+	}
+
+	// translate json to cnp object
+	cnp := &cilium_v2.CiliumNetworkPolicy{}
+	err = json.Unmarshal(jsonData, cnp)
+	if err != nil {
+		return nil, err
+	}
+
+	return cnp, nil
+}
+
+func getLabels(fileName string, cnp *cilium_v2.CiliumNetworkPolicy) labels.LabelArray {
+	labelsArr := labels.LabelArray{
+		labels.NewLabel("filename", fileName, labels.LabelSourceDirectory),
+	}
+
+	ns := cnp.ObjectMeta.Namespace
+	// For clusterwide policy namespace will be empty.
+	if ns != "" {
+		nsLabel := labels.NewLabel(k8sConst.PolicyLabelNamespace, ns, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, nsLabel)
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	} else {
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumClusterwideNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	}
+
+	return labelsArr
+}
+
+// read cilium network policy yaml file and convert to policy object and
+// add rules to policy engine.
+func (p *policyWatcher) addToPolicyEngine(cnpFilePath string) error {
+	// read from file and convert to cnp object
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	cnp.ObjectMeta.Name = filepath.Base(cnpFilePath)
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		cnp.ObjectMeta.Namespace,
+		cnp.ObjectMeta.Name,
+	)
+
+	// convert to rules
+	rules, err := cnp.Parse()
+	if err != nil {
+		return err
+	}
+
+	//update labels
+	lbls := getLabels(cnp.ObjectMeta.Name, cnp)
+	for _, r := range rules {
+		r.Labels = lbls
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceByResource: true,
+		Source:            source.Directory,
+		Resource:          resourceID,
+	})
+
+	if err == nil {
+		p.fileNameToCnpCache[cnp.ObjectMeta.Name] = cnp
+	}
+
+	return err
+}
+
+func (p *policyWatcher) deleteFromPolicyEngine(cnpFilePath string) error {
+	fileName := filepath.Base(cnpFilePath)
+	cnp := p.fileNameToCnpCache[fileName]
+	if cnp == nil {
+		p.log.WithField("file", fileName).Warn("map entry doesn't exist")
```

Comment:
This (and the other log messages here) could be slightly more descriptive. I also would use level=error here to make sure CI fails if we ever hit this case (which we really shouldn't)

```suggestion
		p.log.WithField("file", fileName).Error("BUG: Policy deletion request for file which was never added")
```

Reply:
@gandro so if CNP file is not valid, add will fail and map will not have entry for that file. When user deletes invalid file, map would not have entry for that file and this will be shown as error log but its not an error. I'm more inclined to warn message.. let me know what you think